netbsd/hid.c

```diff
@@ -0,0 +1,1121 @@
+	if (device_handle < 0) {
+		register_device_error_format(dev, "unsupported report id: %hhu", *data);
+		return -1;
+	}
+
+	length--;
+	data++;
+
+	if (length > sizeof(ucr.ucr_data)) {
+		length = sizeof(ucr.ucr_data);
+	}
+
+	ucr.ucr_report = report;
+
+	res = ioctl(device_handle, USB_GET_REPORT, &ucr);
+	if (res == -1) {
+		register_device_error_format(dev, "ioctl (USB_GET_REPORT): %s", strerror(errno));
+		return -1;
+	}
+
+	memcpy(data, ucr.ucr_data, length);
+
+	return (int) (length + 1);
+}
+
+int HID_API_EXPORT HID_API_CALL hid_init(void)
+{
+	const char *locale;
+
+	/* indicate no error */
+	register_global_error(NULL);
+
+	/* Set the locale if it's not set. */
+	locale = setlocale(LC_CTYPE, NULL);
+	if (!locale)
+		setlocale(LC_CTYPE, "");
+
+	return 0;
+}
+
+int HID_API_EXPORT HID_API_CALL hid_exit(void)
+{
+	/* Free global error message */
+	register_global_error(NULL);
+
+	return 0;
+}
+
+struct hid_device_info HID_API_EXPORT * HID_API_CALL hid_enumerate(unsigned short vendor_id, unsigned short product_id)
+{
+	int res;
+	int drvctl;
+	char arr[256][16];
+	size_t len;
+	struct hid_enumerate_data hed;
+
+	res = hid_init();
+	if (res == -1)
+		return NULL;
+
+	drvctl = open(DRVCTLDEV, O_RDONLY | O_CLOEXEC);
+	if (drvctl == -1) {
+		register_global_error_format("failed to open drvctl: %s", strerror(errno));
+		return NULL;
+	}
+
+	len = 0;
+	walk_device_tree(drvctl, "", 0, arr, &len, is_usb_controller);
+
+	hed.root = NULL;
+	hed.end = NULL;
+	hed.drvctl = drvctl;
+	hed.vendor_id = vendor_id;
+	hed.product_id = product_id;
+
+	for (size_t i = 0; i < len; i++) {
+		char devpath[32];
+		int bus;
+
+		strlcpy(devpath, "/dev/", sizeof(devpath));
+		strlcat(devpath, arr[i], sizeof(devpath));
+
+		bus = open(devpath, O_RDONLY | O_CLOEXEC);
+		if (bus == -1)
+			continue;
+
+		enumerate_usb_devices(bus, 0, hid_enumerate_callback, &hed);
+
+		close(bus);
+	}
+
+	close(drvctl);
+
+	return hed.root;
+}
+
+void HID_API_EXPORT HID_API_CALL hid_free_enumeration(struct hid_device_info *devs)
+{
+	while (devs) {
+		struct hid_device_info *next = devs->next;
+		free(devs->path);
+		free(devs->serial_number);
+		free(devs->manufacturer_string);
+		free(devs->product_string);
+		free(devs);
+		devs = next;
+	}
+}
+
+HID_API_EXPORT hid_device * HID_API_CALL hid_open(unsigned short vendor_id, unsigned short product_id, const wchar_t *serial_number)
+{
+	struct hid_device_info *devs;
+	struct hid_device_info *dev;
+	char path[16];
+
+	devs = hid_enumerate(vendor_id, product_id);
+	if (!devs)
+		return NULL;
+
+	*path = '\0';
+
+	for (dev = devs; dev; dev = dev->next) {
+		if (dev->vendor_id != vendor_id)
+			continue;
+
+		if (dev->product_id != product_id)
+			continue;
+
+		if (serial_number && wcscmp(dev->serial_number, serial_number))
+			continue;
+
+		strlcpy(path, dev->path, sizeof(path));
+
+		break;
+	}
+
+	hid_free_enumeration(devs);
+
+	if (*path == '\0') {
+		register_global_error("Device with requested VID/PID/(SerialNumber) not found");
```

Comment:
NOTE: global error never clears
if the previous call has failed, and the next one didn't - the global error still going to point to the previous failure

Reply:
Resolved by clearing global error before `hid_hid_open_path` returns success. All other exit points set a new error.